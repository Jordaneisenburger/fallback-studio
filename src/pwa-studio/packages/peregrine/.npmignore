.eslintrc.js
/lib/**/__{docs,helpers,mocks,tests}__/**
/scripts
