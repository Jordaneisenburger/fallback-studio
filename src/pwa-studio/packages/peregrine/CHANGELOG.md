# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.5.0"></a>
# 0.5.0 (2018-08-13)


### Features

* multicasting REST API client ([#164](https://github.com/magento/pwa-studio/issues/164)) ([2852e14](https://github.com/magento/pwa-studio/commit/2852e14)), closes [#140](https://github.com/magento/pwa-studio/issues/140)
* Simulators for async testing ([#24](https://github.com/magento/pwa-studio/issues/24)) ([a380da9](https://github.com/magento/pwa-studio/commit/a380da9))
