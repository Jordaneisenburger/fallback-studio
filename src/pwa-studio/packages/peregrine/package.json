{
  "name": "@magento/peregrine",
  "version": "2.1.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "dist/index.js",
  "scripts": {
    "build": "concurrently --raw yarn:build:cjs yarn:build:esm",
    "build:cjs": "BABEL_ENV=production babel src --out-dir dist --root-mode 'upward' --source-maps",
    "build:dev": "yarn run build",
    "build:esm": "BABEL_ENV=development babel src --out-dir esm --root-mode 'upward' --source-maps",
    "clean": "rimraf dist esm",
    "prepublishOnly": "yarn run clean && yarn run build",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "storybook:build": "build-storybook -c .storybook -o storybook-dist",
    "watch": "yarn run build --watch"
  },
  "repository": "github:magento-research/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento-research/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine#readme",
  "devDependencies": {
    "@babel/cli": "~7.2.3",
    "@babel/core": "~7.3.4",
    "@babel/plugin-proposal-class-properties": "~7.3.4",
    "@babel/plugin-proposal-object-rest-spread": "~7.3.4",
    "@babel/plugin-syntax-dynamic-import": "~7.2.0",
    "@babel/plugin-syntax-jsx": "~7.2.0",
    "@babel/plugin-transform-react-jsx": "~7.3.0",
    "@babel/plugin-transform-runtime": "~7.4.4",
    "@babel/runtime": "~7.4.2",
    "@storybook/react": "~4.1.13",
    "concurrently": "~4.1.0",
    "intl": "~1.2.5",
    "react": "~16.8.3",
    "react-apollo": "~2.5.3",
    "react-dom": "~16.8.3",
    "react-router-dom": "~5.0.0",
    "react-test-renderer": "~16.8.6",
    "rimraf": "~2.6.3"
  },
  "peerDependencies": {
    "@babel/runtime": "^7.4.2",
    "react": "^16.8.0",
    "react-apollo": "^2.5.3",
    "react-dom": "^16.8.0",
    "react-router-dom": "~5.0.0"
  },
  "engines": {
    "node": "8.10.0 || >=10.7.0",
    "yarn": ">=1.12.0"
  },
  "module": "esm/index.js",
  "jsnext:main": "src/index.js",
  "es2015": "src/index.js",
  "sideEffects": false
}
