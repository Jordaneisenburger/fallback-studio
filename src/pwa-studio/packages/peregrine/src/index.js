import * as RestApi from './RestApi';
import * as Util from './util';

// hooks
export { useApolloContext } from './hooks/useApolloContext';
export { useEventListener } from './hooks/useEventListener';
export { useDropdown } from './hooks/useDropdown';
export { useQuery } from './hooks/useQuery';
export { useQueryResult } from './hooks/useQueryResult';
export { useSearchParam } from './hooks/useSearchParam';
export {
    WindowSizeContextProvider,
    useWindowSize
} from './hooks/useWindowSize';

// components
export { default as ContainerChild } from './ContainerChild';
export { default as List, Items, Item } from './List';
export { default as Page } from './Page';
export { default as Price } from './Price';
export { default as Router } from './Router';

// misc
export { RestApi };
export { Util };
export { default as createTestInstance } from './util/createTestInstance';
