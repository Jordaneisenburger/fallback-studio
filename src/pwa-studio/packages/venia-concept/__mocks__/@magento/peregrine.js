import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/lib');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useApolloContext = jest.fn(peregrine.useApolloContext);
const useEventListener = jest.fn(peregrine.useEventListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useQuery = jest.fn(peregrine.useQuery);
const useQueryResult = jest.fn(peregrine.useQueryResult);
const useSearchParam = jest.fn(peregrine.useSearchParam);
const useWindowSize = jest.fn(peregrine.useWindowSize);
const useToasts = jest.fn(peregrine.useToasts);
// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useApolloContext,
    useEventListener,
    useDropdown,
    useQuery,
    useQueryResult,
    useSearchParam,
    useWindowSize,

    useToasts
};
