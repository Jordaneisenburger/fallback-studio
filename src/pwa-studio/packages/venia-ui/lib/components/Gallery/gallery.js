import React, { useMemo } from 'react';
import { string, shape, array } from 'prop-types';

import { mergeClasses } from '../../classify';
import GalleryItem from './item';
import defaultClasses from './gallery.css';

// map Magento 2.3.1 schema changes to Venia 2.0.0 proptype shape to maintain backwards compatibility
const mapGalleryItem = item => {
    const { small_image } = item;
    return {
        ...item,
        small_image:
            typeof small_image === 'object' ? small_image.url : small_image
    };
};

/**
 * Renders a Gallery of items. If items is an array of nulls Gallery will render
 * a placeholder item for each.
 *
 * @params {Array} props.items an array of items to render
 */
const Gallery = props => {
    const classes = mergeClasses(defaultClasses, props.classes);

    const { items } = props;

    const galleryItems = useMemo(
        () =>
            items.map((item, index) => {
                if (item === null) {
                    return <GalleryItem key={index} />;
                }
                return <GalleryItem key={index} item={mapGalleryItem(item)} />;
            }),
        [items]
    );

    return (
        <div className={classes.root}>
            <div className={classes.items}>{galleryItems}</div>
        </div>
    );
};

Gallery.propTypes = {
    classes: shape({
        filters: string,
        items: string,
        pagination: string,
        root: string
    }),
    items: array.isRequired
};

export default Gallery;
