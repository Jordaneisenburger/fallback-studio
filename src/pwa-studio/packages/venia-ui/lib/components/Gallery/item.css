.root {
}

.images {
    display: grid;
    grid-template-areas: 'main';
    margin-bottom: 0.65rem;
}

.imageContainer {
    grid-area: main;
}

.image {
    display: block;
    height: auto;
    opacity: 1;
    transition-duration: 512ms;
    transition-property: opacity, visibility;
    transition-timing-function: ease-out;
    visibility: visible;
    width: 100%;
}

.imagePlaceholder {
    composes: image;
    background-color: rgb(var(--venia-grey));
}

.name,
.price {
    font-size: 0.875rem;
    line-height: 1rem;
    margin: 0.5rem 0;
    min-height: 1rem;
}

/* state: pending */

.root_pending {
    composes: root;
}

.image_pending {
    composes: image;
    background-color: rgb(var(--venia-grey));
}

.images_pending {
    composes: images;
}

.name_pending {
    composes: name;
    background-color: rgb(var(--venia-grey));
}

.price_pending {
    composes: price;
    background-color: rgb(var(--venia-grey));
    width: 3rem;
}
