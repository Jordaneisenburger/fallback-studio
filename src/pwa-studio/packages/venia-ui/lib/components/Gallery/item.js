import React from 'react';
import { string, number, shape } from 'prop-types';
import { Link, resourceUrl } from '@magento/venia-drivers';
import { Price } from '@magento/peregrine';
import { transparentPlaceholder } from '@magento/peregrine/lib/util/images';
import { UNCONSTRAINED_SIZE_KEY } from '@magento/peregrine/lib/talons/Image/useImage';

import { mergeClasses } from '../../classify';
import Image from '../Image';
import defaultClasses from './item.css';

// The placeholder image is 4:5, so we should make sure to size our product
// images appropriately.
const IMAGE_WIDTH = 300;
const IMAGE_HEIGHT = 375;

// Gallery switches from two columns to three at 640px.
const IMAGE_WIDTHS = new Map()
    .set(640, IMAGE_WIDTH)
    .set(UNCONSTRAINED_SIZE_KEY, 840);

// TODO: get productUrlSuffix from graphql when it is ready
const productUrlSuffix = '.html';

const ItemPlaceholder = ({ classes }) => (
    <div className={classes.root_pending}>
        <div className={classes.images_pending}>
            <Image
                alt="Placeholder for gallery item image"
                classes={{
                    image: classes.image_pending,
                    root: classes.imageContainer
                }}
                src={transparentPlaceholder}
            />
        </div>
        <div className={classes.name_pending} />
        <div className={classes.price_pending} />
    </div>
);

const GalleryItem = props => {
    const { item } = props;

    const classes = mergeClasses(defaultClasses, props.classes);

    if (!item) {
        return <ItemPlaceholder classes={classes} />;
    }

    const { name, price, small_image, url_key } = item;
    const productLink = resourceUrl(`/${url_key}${productUrlSuffix}`);

    return (
        <div className={classes.root}>
            <Link to={productLink} className={classes.images}>
                <Image
                    alt={name}
                    classes={{
                        image: classes.image,
                        root: classes.imageContainer
                    }}
                    height={IMAGE_HEIGHT}
                    resource={small_image}
                    widths={IMAGE_WIDTHS}
                />
            </Link>
            <Link to={productLink} className={classes.name}>
                <span>{name}</span>
            </Link>
            <div className={classes.price}>
                <Price
                    value={price.regularPrice.amount.value}
                    currencyCode={price.regularPrice.amount.currency}
                />
            </div>
        </div>
    );
};

GalleryItem.propTypes = {
    classes: shape({
        image: string,
        imageContainer: string,
        imagePlaceholder: string,
        image_pending: string,
        images: string,
        images_pending: string,
        name: string,
        name_pending: string,
        price: string,
        price_pending: string,
        root: string,
        root_pending: string
    }),
    item: shape({
        id: number.isRequired,
        name: string.isRequired,
        small_image: string.isRequired,
        url_key: string.isRequired,
        price: shape({
            regularPrice: shape({
                amount: shape({
                    value: number.isRequired,
                    currency: string.isRequired
                }).isRequired
            }).isRequired
        }).isRequired
    })
};

export default GalleryItem;
