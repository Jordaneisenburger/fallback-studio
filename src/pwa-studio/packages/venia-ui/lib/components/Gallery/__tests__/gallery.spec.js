import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import Gallery from '../gallery';

jest.mock('@magento/venia-drivers', () => ({
    Link: ({ children }) => children,
    resourceUrl: () => 'a.url'
}));
jest.mock('@magento/peregrine/lib/talons/Image/useImage', () => {
    return {
        useImage: () => ({
            handleError: jest.fn(),
            handleImageLoad: jest.fn(),
            hasError: false,
            isLoaded: true,
            resourceWidth: 100
        })
    };
});
jest.mock('../../../classify');

const classes = { root: 'foo' };
const items = [
    {
        id: 1,
        name: 'Test Product 1',
        small_image: {
            url: '/test/product/1.png'
        },
        price: {
            regularPrice: {
                amount: {
                    value: 100,
                    currency: 'USD'
                }
            }
        },
        url_key: 'test-product1'
    },
    {
        id: 2,
        name: 'Test Product 2',
        // Magento 2.3.0 schema for testing backwards compatibility
        small_image: '/test/product/2.png',
        price: {
            regularPrice: {
                amount: {
                    value: 100,
                    currency: 'USD'
                }
            }
        },
        url_key: 'test-product2'
    }
];

test('renders if `items` is an empty array', () => {
    const wrapper = createTestInstance(
        <Gallery classes={classes} items={[]} />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
});

test('renders if `items` is an array of objects', () => {
    const wrapper = createTestInstance(
        <Gallery classes={classes} items={items} />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
});
