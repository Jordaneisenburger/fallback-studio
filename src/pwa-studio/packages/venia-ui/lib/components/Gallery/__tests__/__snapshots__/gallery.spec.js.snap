// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders if \`items\` is an array of objects 1`] = `
<div
  className="root"
>
  <div
    className="items"
  >
    <div
      className="root"
    >
      <div
        className="root container"
      >
        <img
          alt="Test Product 1"
          className="image placeholder_layoutOnly"
          height={375}
          loading="eager"
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAAFCAQAAADIpIVQAAAADklEQVR42mNkgAJGIhgAALQABsHyMOcAAAAASUVORK5CYII="
          width={100}
        />
        <img
          alt="Test Product 1"
          className="image loaded"
          height={375}
          loading="lazy"
          onError={[MockFunction]}
          onLoad={[MockFunction]}
          sizes="(max-width: 640px) 300px, 840px"
          src="a.url"
          srcSet="a.url 40w,
a.url 80w,
a.url 160w,
a.url 320w,
a.url 640w,
a.url 960w,
a.url 1280w,
a.url 1600w,
a.url 2560w"
          width={100}
        />
      </div>
      <span>
        Test Product 1
      </span>
      <div
        className="price"
      >
        <div />
      </div>
    </div>
    <div
      className="root"
    >
      <div
        className="root container"
      >
        <img
          alt="Test Product 2"
          className="image placeholder_layoutOnly"
          height={375}
          loading="eager"
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAAFCAQAAADIpIVQAAAADklEQVR42mNkgAJGIhgAALQABsHyMOcAAAAASUVORK5CYII="
          width={100}
        />
        <img
          alt="Test Product 2"
          className="image loaded"
          height={375}
          loading="lazy"
          onError={[MockFunction]}
          onLoad={[MockFunction]}
          sizes="(max-width: 640px) 300px, 840px"
          src="a.url"
          srcSet="a.url 40w,
a.url 80w,
a.url 160w,
a.url 320w,
a.url 640w,
a.url 960w,
a.url 1280w,
a.url 1600w,
a.url 2560w"
          width={100}
        />
      </div>
      <span>
        Test Product 2
      </span>
      <div
        className="price"
      >
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`renders if \`items\` is an empty array 1`] = `
<div
  className="root"
>
  <div
    className="items"
  />
</div>
`;
